Competition Task

JigyashaSharma opened the pull request:
This PR consist of commits for the competition task.

Employer profile page
- Task 1 :  Add last name to the primary contact details.

-Task2:  Allow users to edit company contact details by clicking on the edit button

-Task3: Display the user's full name on primary contact details (for read only display)

Manage Job page
- Task 4 :  Display jobs as a list of cards. Add a new job by clicking on ‘Post a Job’ button.
- Task5:    Update a job, Close a job. ManageJob page.

Last commit: Validations for Employer profile page.
This commit I added the validations for the Employer profile page which I missed in earlier commits.

Added the screenshot.

JigyashaSharma committed:
Task1: Add last name to the primary contact details.

Added the last name in Employer Primary Contact Details.

JigyashaSharma committed:
Task2: Allow users to edit company contact details by clicking on the edit button.

JigyashaSharma committed:
Task3: Display the user's full name on primary contact details (for read only display)

JigyashaSharma committed:
Task 4: Add a new job by clicking on ‘Post a Job’ button.

1. Add a new job by clicking on ‘Post a Job’ button
2. Check the Mongo DB database for the newly inserted record (under the Job Collection)

JigyashaSharma committed:
Task5: ManageJob page.

ManageJob.jsx and ManageJobsFilters.jsx got committed in previous commit. Please refer it there.
Functionalities given:
1. Display No Jobs Found.
2. Display with jobs.
3. Filters
4. Sorting by date
5. Pagination
6. Close the Job
7. Edit the Job
8. Copy the Job

JigyashaSharma committed:
Task: Validations for Employer profile page.

Adding the validation for Primary and Company details edit.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/CreateJob/CreateJob.jsx:
```diff
@@ -91,13 +137,59 @@ export default class CreateJob extends React.Component {
+    }
+
+    //Add the required field to requiredFields array.
+    validateField() {
+        const missingFieldsTemp = requiredFields.reduce((acc, field) => {
+            const value = field.name.split('.').reduce((obj, prop) => obj && obj[prop], this.state.jobData);
+            if (!value || (Array.isArray(value) && value.length === 0)) {
+                return { ...acc, [field.name]: field.message };
+            }
+            return acc;
+        }, {});
+
+        if (Object.keys(missingFieldsTemp).length > 0) {
+            this.setState({ formErrors: missingFieldsTemp });
+            return false;
+        } else {
+            return true;
+        }
```
This is good :)

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/CreateJob/JobSummary.jsx:
```diff
@@ -22,6 +22,7 @@ export class JobSummary extends React.Component {
                     onChange={this.updateField}
                 >
                 </textarea>
+                {this.props.isError ? <div className="ui basic red pointing prompt label transition visible">{this.props.errorMessage}</div> : null}
```
This setup seems present on multiple files and they have the same classNames. You can actually create a component for this error message

kestermvp replied:
Have you seen this @JigyashaSharma?

JigyashaSharma replied:
Yes Kester, working on it.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/CloseJobConfirmation.jsx:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    handleCloseCancel = () => {
+        console.log("handleCloseCancel CloseJobConfirmation");
+        this.props.setCloseConfirm(false);
+        this.props.setCloseJobId(null);
+    };
+
+    handleCloseConfirm = () => {
+         console.log("handleCloseConfirm CloseJobConfirmation");
+        this.props.handleClosejob(this.props.closeJobId);
+        this.props.setCloseJobId(null);
+        this.props.setCloseConfirm(false);
+    };
+
+    render() {
+        { console.log("close modal", this.props.open) }
```
Can you please remove the console.logs

kestermvp replied:
Have you seen this as well @JigyashaSharma?

JigyashaSharma replied:
Yes Kester, working on it.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/DisplayJobs.jsx:
```diff
@@ -0,0 +1,125 @@
+        return (
+            <Container>
+                <Card.Group>
+                    {pagedJobs.map((job) => (
```
Upon checking the source of **paginatedJobs** it seems that the main value is using an empty array which means **.map** in here. However, if you want to have more security on your code and to make sure that nothing will break, instead of directly mapping it here, you can create a separate function and then, you can validate first if **pagedJobs** is an actual array and contains some items. This way, you can be able to make sure that your code will be able to work based on what you expect.

kestermvp replied:
Hi @JigyashaSharma, have you seen this comment?

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/DisplayJobs.jsx:
```diff
@@ -0,0 +1,125 @@
+            <Container>
+                <Card.Group>
+                    {pagedJobs.map((job) => (
+                        <Card key={job.id} style={{ marginRight: '10px' }} >
+                                    <Card.Content className="eight wide computer eight wide tablet sixteen wide mobile column">
+                                        <Card.Header>{job.title}</Card.Header>
+                                        {/* ... your card content ... */}
+                                        <a class="ui black right ribbon label">
+                                            <Icon name='user' size='small' />
+                                            {job.noOfSuggestions}
+                                        </a>
+                                        <Card.Meta>{job.location.city}, &nbsp; {job.location.country}</Card.Meta>
+                                        <Card.Description>{job.summary}</Card.Description>
+                                    </Card.Content>
+                            
+                            {/* Line at the bottom of the card */}
+                            <div className="ui divider"></div> 
+                            <div className="ui grid">
+                                <div className="three wide column" style={{ marginLeft: '5px', marginBottom: '10px'} }>
```
I can see some inline styling in your changes. Though working properly, it will be tedious and hard to maintain. Instead of this, you can:

1. Maximise the third-party styling library if you have any
2. Create a new custom styling file and import it to any files that you need

JigyashaSharma replied:
I tried to put the styles in file and import it. But the style was not reflecting, not sure what was wrong. I thought I will ask this to you after holidays. For now I put those as inline. 
I will try to Maximise the third-party library.

kestermvp replied:
Hey @JigyashaSharma, you can create a separate CSS file. For example :
`
.sampleStyle {
 yourstylehere
}
`

and then, you can import this like:
`import styles from './NewCSS.css';`

then you can use it into something like
`<div className={styles.sampleStyle}>`

Let me know if this will work or not on your end

JigyashaSharma replied:
Hi @kestermvp ,

I tried this but I am getting the : TypeError: Cannot read properties of undefined (reading 'cardMargin')

JoStyle.module.css file:
.cardMargin {
    margin-right: 10px;
}
.expireJobMargin {
    margin-left: 5px;
    margin-bottom: 10px;
}

DisplayJobs.jsx
import styles from "./JobStyle.module.css";
`<Card key={job.id} className={styles.cardMargin} >`
-----------------------------------------------------------------
I tried to print the console.log("styles check: ",styles);  
styles shows undefined.
----------------------------------------------------
I also tried the regular css, there was no error but styles were not proper.

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/JobEditNavigator.jsx:
```diff
@@ -0,0 +1,28 @@
+import DisplayJobs from './DisplayJobs.jsx';
+
+const JobEditNavigator = ( props ) => {
+    const navigate = useNavigate(); // Get navigate function using useNavigate hook
+
+    const handleEditClick = (jobId) => {
+        // Navigate to createJob page with jobId as a URL parameter
+        navigate(`/EditJob/${jobId}`);
+    };
+
+    const handleCopyClick = (jobId) => {
+        navigate(`/PostJob/${jobId}`);
+    };
```
What will happen if **jobId** contains an empty value?

kestermvp replied:
Hi @JigyashaSharma, have you seen this comment?

kestermvp commented on App/Talent.App.WebApp/wwwroot/js/react/ReactScripts/Employer/ManageJob/ManageJob.jsx:
```diff
@@ -72,12 +152,117 @@ export default class ManageJob extends React.Component {
+    };
+
+    handleClosejob = (jobId) => {
+        const link = 'http://localhost:51689/listing/listing/CloseJob';
```
Is there any reason why you are using local endpoint here?

JigyashaSharma replied:
In the production code its Azure endpoint. I didn't replace the localhost endpoints for the PR since it would add lot of these changes and files. Shall I replace all the localhost endpoints here as well?

kestermvp replied:
Hi @JigyashaSharma, this is fine for now. But, please keep in mind that in the actual project, we need to use the production endpoint **or** we can use environment variables to distinguish different environments such as development, staging, or production.

For our PRs, we should always make sure that the content should reflect what we can from the production level, because these PRs will be going to be merged on the master branch which is always the source of truth

JigyashaSharma replied:
Done changes.

JigyashaSharma committed:
Addressing Review Comments

1. Create separate component for Error Message.
2. Removed console logs in CloseJobConfirmation.jsx.
3. Handled empty jobId in JobEditNavigator.jsx.
4. Committed the local endpoints with Production endpoints.

JigyashaSharma committed:
Removing inline css and adding the Module css.

JigyashaSharma committed:
Removing console log used for debugging

kestermvp reviewed (approved):
Hi @JigyashaSharma, your recent changes look good :)
